src/query/api/v1/handler/prometheus/handleroptions/tag_options.go

```diff
@@ -111,3 +111,81 @@ type StringTagOptions struct {
 	Restrict []StringMatch `json:"match"`
 	Strip    []string      `json:"strip"`
 }
+
+// MapTagsOptions representations mutations to be applied to all timeseries in a
+// write request.
+type MapTagsOptions struct {
+	TagMappers []TagMapper `json:"tagMappers"`
+}
+
+// TagMapper represents one of a variety of tag mapping operations.
+type TagMapper struct {
+	Append  AppendOp  `json:"append,omitEmpty"`
+	Drop    DropOp    `json:"drop,omitEmpty"`
+	Replace ReplaceOp `json:"replace,omitEmpty"`
+}
+
+// Validate ensures the mapper is valid.
+func (t TagMapper) Validate() error {
+	numOps := 0
+	if !t.Append.IsEmpty() {
+		numOps++
+	}
+
+	if !t.Drop.IsEmpty() {
+		numOps++
+	}
+
+	if !t.Replace.IsEmpty() {
+		numOps++
+	}
+
+	if numOps == 1 {
+		return nil
+	}
+
+	return fmt.Errorf("must specify one operation per tag mapper (got %d)", numOps)
+}
+
+// AppendOp with value tag="foo" and value="bar" will unconditionally add
```

Comment:
nit: maybe change the name from `append` to `write` or something? Drop and Replace seem to describe their operations well enough, but this is more of an addOrUpdate than an append?